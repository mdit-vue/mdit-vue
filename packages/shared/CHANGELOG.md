# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

## [3.0.2](https://github.com/mdit-vue/mdit-vue/compare/v3.0.1...v3.0.2) (2025-08-11)

**Note:** Version bump only for package @mdit-vue/shared

## [3.0.1](https://github.com/mdit-vue/mdit-vue/compare/v3.0.0...v3.0.1) (2025-08-11)

**Note:** Version bump only for package @mdit-vue/shared

# [3.0.0](https://github.com/mdit-vue/mdit-vue/compare/v2.1.4...v3.0.0) (2025-08-11)

### Build System

- drop node 18 ([a4775a6](https://github.com/mdit-vue/mdit-vue/commit/a4775a6f3017ee1065075352429cde5587df8c0f))

### BREAKING CHANGES

- drop node 18

## [2.1.4](https://github.com/mdit-vue/mdit-vue/compare/v2.1.3...v2.1.4) (2025-04-10)

**Note:** Version bump only for package @mdit-vue/shared

## [2.1.3](https://github.com/mdit-vue/mdit-vue/compare/v2.1.2...v2.1.3) (2024-05-13)

**Note:** Version bump only for package @mdit-vue/shared

## [2.1.2](https://github.com/mdit-vue/mdit-vue/compare/v2.1.1...v2.1.2) (2024-04-15)

### Bug Fixes

- sync markdown-it types updates ([#9](https://github.com/mdit-vue/mdit-vue/issues/9)) ([f6768b1](https://github.com/mdit-vue/mdit-vue/commit/f6768b10e0cd06843886448065369a1c3aa143bd))

# [2.1.0](https://github.com/mdit-vue/mdit-vue/compare/v2.0.0...v2.1.0) (2024-04-08)

### Features

- upgrade markdown-it types to v14 ([6839be8](https://github.com/mdit-vue/mdit-vue/commit/6839be8c090c9c1faa73c17a048a37cdf53b4625))

# [2.0.0](https://github.com/mdit-vue/mdit-vue/compare/v1.0.0...v2.0.0) (2023-12-08)

### Features

- bump to markdown-it 14 ([522950f](https://github.com/mdit-vue/mdit-vue/commit/522950f09cdbff37a26dbd5fccbbd57f931b3727))

### BREAKING CHANGES

- bump to markdown-it 14

# [1.0.0](https://github.com/mdit-vue/mdit-vue/compare/v0.12.1...v1.0.0) (2023-09-26)

**Note:** Version bump only for package @mdit-vue/shared

## [0.12.1](https://github.com/mdit-vue/mdit-vue/compare/v0.12.0...v0.12.1) (2023-08-24)

**Note:** Version bump only for package @mdit-vue/shared

# [0.12.0](https://github.com/mdit-vue/mdit-vue/compare/v0.11.2...v0.12.0) (2023-02-10)

### Features

- **shared:** add shouldAllowNested option for resolveHeadersFromTokens ([9363b92](https://github.com/mdit-vue/mdit-vue/commit/9363b92c3ac3efbe47c1c58300ff10bbe93d35b3))

## [0.11.2](https://github.com/mdit-vue/mdit-vue/compare/v0.11.1...v0.11.2) (2022-12-23)

### Bug Fixes

- **shared:** avoid resolve headers in nested blocks ([ce146bd](https://github.com/mdit-vue/mdit-vue/commit/ce146bdb3f3c84320a9f7a0a4fac6f3f67bd67f8))

# [0.11.0](https://github.com/mdit-vue/mdit-vue/compare/v0.10.0...v0.11.0) (2022-08-25)

**Note:** Version bump only for package @mdit-vue/shared

# [0.10.0](https://github.com/mdit-vue/mdit-vue/compare/v0.9.4...v0.10.0) (2022-08-23)

### Features

- **plugin-headers:** resolve link property for headers ([da2954f](https://github.com/mdit-vue/mdit-vue/commit/da2954fbf89d40df2e9015784c5e66f8dd97da2a))

## [0.9.2](https://github.com/mdit-vue/mdit-vue/compare/v0.9.1...v0.9.2) (2022-08-18)

**Note:** Version bump only for package @mdit-vue/shared

## [0.9.1](https://github.com/mdit-vue/mdit-vue/compare/v0.9.0...v0.9.1) (2022-07-26)

**Note:** Version bump only for package @mdit-vue/shared

# [0.9.0](https://github.com/mdit-vue/mdit-vue/compare/v0.8.1...v0.9.0) (2022-07-26)

**Note:** Version bump only for package @mdit-vue/shared

# [0.8.0](https://github.com/mdit-vue/mdit-vue/compare/v0.7.1...v0.8.0) (2022-07-22)

**Note:** Version bump only for package @mdit-vue/shared

## [0.7.1](https://github.com/mdit-vue/mdit-vue/compare/v0.7.0...v0.7.1) (2022-07-20)

**Note:** Version bump only for package @mdit-vue/shared

# [0.7.0](https://github.com/mdit-vue/mdit-vue/compare/v0.6.0...v0.7.0) (2022-07-18)

**Note:** Version bump only for package @mdit-vue/shared

# [0.6.0](https://github.com/mdit-vue/mdit-vue/compare/v0.5.0...v0.6.0) (2022-07-11)

### Features

- **types:** extract types from shared package ([5fc56ce](https://github.com/mdit-vue/mdit-vue/commit/5fc56ce6439159584e09e13ca9a6e87abe8f4389))

# [0.5.0](https://github.com/mdit-vue/mdit-vue/compare/v0.4.0...v0.5.0) (2022-07-06)

**Note:** Version bump only for package @mdit-vue/shared

# [0.4.0](https://github.com/mdit-vue/mdit-vue/compare/v0.3.1...v0.4.0) (2022-07-02)

**Note:** Version bump only for package @mdit-vue/shared

# [0.3.0](https://github.com/mdit-vue/mdit-vue/compare/v0.2.0...v0.3.0) (2022-06-25)

### Features

- **plugin-toc:** add toc plugin ([e64cc2c](https://github.com/mdit-vue/mdit-vue/commit/e64cc2ce3bbc89999d373f8a00fb3a049d6af58f))
- **shared:** add html escape and unescape utils ([e1103c2](https://github.com/mdit-vue/mdit-vue/commit/e1103c21998aec77c67cdefd3cbb4994c127124c))

# [0.2.0](https://github.com/mdit-vue/mdit-vue/compare/v0.1.1...v0.2.0) (2022-06-23)

### Features

- **shared:** add shared package ([9e36bdc](https://github.com/mdit-vue/mdit-vue/commit/9e36bdce8635ae048e8b0f2cc2d02cc0448ad09f))
