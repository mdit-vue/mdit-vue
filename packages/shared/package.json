{
  "name": "@mdit-vue/shared",
  "version": "2.1.0",
  "description": "Shared utils and types of mdit-vue",
  "keywords": [
    "markdown-it",
    "vue",
    "utils"
  ],
  "homepage": "https://github.com/mdit-vue",
  "bugs": {
    "url": "https://github.com/mdit-vue/mdit-vue/issues"
  },
  "repository": {
    "type": "git",
    "url": "git+https://github.com/mdit-vue/mdit-vue.git"
  },
  "license": "MIT",
  "author": "meteorlxy <meteor.lxy@foxmail.com>",
  "type": "module",
  "exports": {
    ".": {
      "import": {
        "types": "./dist/index.d.mts",
        "default": "./dist/index.mjs"
      },
      "require": {
        "types": "./dist/index.d.cts",
        "default": "./dist/index.cjs"
      }
    }
  },
  "main": "./dist/index.cjs",
  "module": "./dist/index.mjs",
  "types": "./dist/index.d.ts",
  "files": [
    "./dist"
  ],
  "scripts": {
    "build": "unbuild",
    "test": "vitest"
  },
  "dependencies": {
    "@mdit-vue/types": "workspace:*",
    "@types/markdown-it": "^14.0.1",
    "markdown-it": "^14.1.0"
  },
  "publishConfig": {
    "access": "public"
  }
}
