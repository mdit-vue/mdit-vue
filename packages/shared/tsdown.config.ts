import { defineConfig } from 'tsdown';

export default defineConfig({
  dts: true,
  entry: './src/index.ts',
  exports: true,
  fixedExtension: true,
  format: ['cjs', 'esm'],
  outDir: 'dist',
});
