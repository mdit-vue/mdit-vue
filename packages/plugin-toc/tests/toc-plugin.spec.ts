import { slugify } from '@mdit-vue/shared';
import MarkdownIt from 'markdown-it';
import { default as anchorPlugin } from 'markdown-it-anchor';
import { describe, expect, it } from 'vitest';
import { tocPlugin } from '../src/index.js';

const FIXTURES = {
  simpleTree: `\
[[toc]]
# h1
## h2
### h3
#### h4
##### h5
###### h6
`,
  complexTree: `\
[[toc]]
# s1
## s1-1
### s1-1-1
#### s1-1-1-1
### s1-1-2
### s1-1-3
#### s1-1-3-1
## s1-2
## s1-3
### s1-3-2
#### s1-3-2-1
##### s1-3-2-1-1
##### s1-3-2-1-2
`,
  reversedTree: `\
[[toc]]
###### h6
##### h5
#### h4
### h3
## h2
# h1
`,
};

describe('should render toc with default option (h2, h3)', () => {
  const md = MarkdownIt().use(tocPlugin);

  Object.entries(FIXTURES).forEach(([name, source]) => {
    it(name, () => {
      const result = md.render(source);
      expect(result).toMatchSnapshot();
    });
  });
});

describe('should render toc (h1, h2, h3, h4)', () => {
  const md = MarkdownIt().use(tocPlugin, {
    level: [1, 2, 3, 4],
  });

  Object.entries(FIXTURES).forEach(([name, source]) => {
    it(name, () => {
      const result = md.render(source);
      expect(result).toMatchSnapshot();
    });
  });
});

describe('should render toc with RouterLink', () => {
  const md = MarkdownIt().use(tocPlugin, {
    linkTag: 'RouterLink',
  });

  Object.entries(FIXTURES).forEach(([name, source]) => {
    it(name, () => {
      const result = md.render(source);
      expect(result).toMatchSnapshot();
    });
  });
});

describe('should render class name correctly', () => {
  const md = MarkdownIt().use(tocPlugin, {
    // remove default container class
    containerClass: '',
    // add custom list / item / link class
    listClass: 'toc-list',
    itemClass: 'toc-item',
    linkClass: 'toc-link',
  });

  Object.entries(FIXTURES).forEach(([name, source]) => {
    it(name, () => {
      const result = md.render(source);
      expect(result).toMatchSnapshot();
    });
  });
});

describe('should include html elements and should escape texts', () => {
  const md = MarkdownIt({
    html: true,
  })
    .use(anchorPlugin, {
      level: [1, 2, 3, 4, 5, 6],
      slugify,
      permalink: anchorPlugin.permalink.ariaHidden({
        class: 'header-anchor',
        symbol: '#',
        space: true,
        placement: 'before',
      }),
    })
    .use(tocPlugin, { slugify });

  const TEST_CASES: [string, { slug: string; title: string; h2: string }][] = [
    // html element should be kept as is
    [
      `\
[[toc]]
## foo <bar />
`,
      {
        slug: 'foo',
        title: 'foo <bar />',
        h2: 'foo <bar />',
      },
    ],
    // inline code should be escaped
    [
      `\
[[toc]]
## foo <bar /> \`<code />\`
`,
      {
        slug: 'foo-code',
        title: 'foo <bar /> &lt;code /&gt;',
        h2: 'foo <bar /> <code>&lt;code /&gt;</code>',
      },
    ],
    // text should be escaped
    [
      `\
[[toc]]
## foo <bar/> "baz"
`,
      {
        slug: 'foo-baz',
        title: 'foo <bar/> &quot;baz&quot;',
        h2: 'foo <bar/> &quot;baz&quot;',
      },
    ],
    // text should be escaped
    [
      `\
[[toc]]
## < test >
`,
      {
        slug: 'test',
        title: '&lt; test &gt;',
        h2: '&lt; test &gt;',
      },
    ],
  ];

  TEST_CASES.forEach(([source, expected], i) => {
    it(`case ${i}`, () => {
      expect(md.render(source)).toEqual(`\
<nav class="table-of-contents"><ul><li><a href="#${expected.slug}">${expected.title}</a></li></ul></nav>
<h2 id="${expected.slug}" tabindex="-1"><a class="header-anchor" href="#${expected.slug}" aria-hidden="true">#</a> ${expected.h2}</h2>
`);
    });
  });
});

describe('edge cases', () => {
  it('should not terminate the blockquote', () => {
    const md = MarkdownIt().use(tocPlugin);

    const source = `\
> foo
    [[toc]]
`;
    expect(md.render(source)).toEqual(`\
<blockquote>
<p>foo
[[toc]]</p>
</blockquote>
`);
  });
});
