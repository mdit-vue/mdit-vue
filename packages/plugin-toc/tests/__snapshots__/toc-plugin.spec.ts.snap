// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`should render class name correctly > complexTree 1`] = `
"<nav><ul class="toc-list"><li class="toc-item"><a class="toc-link" href="#s1-1">s1-1</a><ul class="toc-list"><li class="toc-item"><a class="toc-link" href="#s1-1-1">s1-1-1</a></li><li class="toc-item"><a class="toc-link" href="#s1-1-2">s1-1-2</a></li><li class="toc-item"><a class="toc-link" href="#s1-1-3">s1-1-3</a></li></ul></li><li class="toc-item"><a class="toc-link" href="#s1-2">s1-2</a></li><li class="toc-item"><a class="toc-link" href="#s1-3">s1-3</a><ul class="toc-list"><li class="toc-item"><a class="toc-link" href="#s1-3-2">s1-3-2</a></li></ul></li></ul></nav>
<h1>s1</h1>
<h2>s1-1</h2>
<h3>s1-1-1</h3>
<h4>s1-1-1-1</h4>
<h3>s1-1-2</h3>
<h3>s1-1-3</h3>
<h4>s1-1-3-1</h4>
<h2>s1-2</h2>
<h2>s1-3</h2>
<h3>s1-3-2</h3>
<h4>s1-3-2-1</h4>
<h5>s1-3-2-1-1</h5>
<h5>s1-3-2-1-2</h5>
"
`;

exports[`should render class name correctly > reversedTree 1`] = `
"<nav><ul class="toc-list"><li class="toc-item"><a class="toc-link" href="#h3">h3</a></li><li class="toc-item"><a class="toc-link" href="#h2">h2</a></li></ul></nav>
<h6>h6</h6>
<h5>h5</h5>
<h4>h4</h4>
<h3>h3</h3>
<h2>h2</h2>
<h1>h1</h1>
"
`;

exports[`should render class name correctly > simpleTree 1`] = `
"<nav><ul class="toc-list"><li class="toc-item"><a class="toc-link" href="#h2">h2</a><ul class="toc-list"><li class="toc-item"><a class="toc-link" href="#h3">h3</a></li></ul></li></ul></nav>
<h1>h1</h1>
<h2>h2</h2>
<h3>h3</h3>
<h4>h4</h4>
<h5>h5</h5>
<h6>h6</h6>
"
`;

exports[`should render toc (h1, h2, h3, h4) > complexTree 1`] = `
"<nav class="table-of-contents"><ul><li><a href="#s1">s1</a><ul><li><a href="#s1-1">s1-1</a><ul><li><a href="#s1-1-1">s1-1-1</a><ul><li><a href="#s1-1-1-1">s1-1-1-1</a></li></ul></li><li><a href="#s1-1-2">s1-1-2</a></li><li><a href="#s1-1-3">s1-1-3</a><ul><li><a href="#s1-1-3-1">s1-1-3-1</a></li></ul></li></ul></li><li><a href="#s1-2">s1-2</a></li><li><a href="#s1-3">s1-3</a><ul><li><a href="#s1-3-2">s1-3-2</a><ul><li><a href="#s1-3-2-1">s1-3-2-1</a></li></ul></li></ul></li></ul></li></ul></nav>
<h1>s1</h1>
<h2>s1-1</h2>
<h3>s1-1-1</h3>
<h4>s1-1-1-1</h4>
<h3>s1-1-2</h3>
<h3>s1-1-3</h3>
<h4>s1-1-3-1</h4>
<h2>s1-2</h2>
<h2>s1-3</h2>
<h3>s1-3-2</h3>
<h4>s1-3-2-1</h4>
<h5>s1-3-2-1-1</h5>
<h5>s1-3-2-1-2</h5>
"
`;

exports[`should render toc (h1, h2, h3, h4) > reversedTree 1`] = `
"<nav class="table-of-contents"><ul><li><a href="#h4">h4</a></li><li><a href="#h3">h3</a></li><li><a href="#h2">h2</a></li><li><a href="#h1">h1</a></li></ul></nav>
<h6>h6</h6>
<h5>h5</h5>
<h4>h4</h4>
<h3>h3</h3>
<h2>h2</h2>
<h1>h1</h1>
"
`;

exports[`should render toc (h1, h2, h3, h4) > simpleTree 1`] = `
"<nav class="table-of-contents"><ul><li><a href="#h1">h1</a><ul><li><a href="#h2">h2</a><ul><li><a href="#h3">h3</a><ul><li><a href="#h4">h4</a></li></ul></li></ul></li></ul></li></ul></nav>
<h1>h1</h1>
<h2>h2</h2>
<h3>h3</h3>
<h4>h4</h4>
<h5>h5</h5>
<h6>h6</h6>
"
`;

exports[`should render toc with RouterLink > complexTree 1`] = `
"<nav class="table-of-contents"><ul><li><RouterLink href="#s1-1">s1-1</RouterLink><ul><li><RouterLink href="#s1-1-1">s1-1-1</RouterLink></li><li><RouterLink href="#s1-1-2">s1-1-2</RouterLink></li><li><RouterLink href="#s1-1-3">s1-1-3</RouterLink></li></ul></li><li><RouterLink href="#s1-2">s1-2</RouterLink></li><li><RouterLink href="#s1-3">s1-3</RouterLink><ul><li><RouterLink href="#s1-3-2">s1-3-2</RouterLink></li></ul></li></ul></nav>
<h1>s1</h1>
<h2>s1-1</h2>
<h3>s1-1-1</h3>
<h4>s1-1-1-1</h4>
<h3>s1-1-2</h3>
<h3>s1-1-3</h3>
<h4>s1-1-3-1</h4>
<h2>s1-2</h2>
<h2>s1-3</h2>
<h3>s1-3-2</h3>
<h4>s1-3-2-1</h4>
<h5>s1-3-2-1-1</h5>
<h5>s1-3-2-1-2</h5>
"
`;

exports[`should render toc with RouterLink > reversedTree 1`] = `
"<nav class="table-of-contents"><ul><li><RouterLink href="#h3">h3</RouterLink></li><li><RouterLink href="#h2">h2</RouterLink></li></ul></nav>
<h6>h6</h6>
<h5>h5</h5>
<h4>h4</h4>
<h3>h3</h3>
<h2>h2</h2>
<h1>h1</h1>
"
`;

exports[`should render toc with RouterLink > simpleTree 1`] = `
"<nav class="table-of-contents"><ul><li><RouterLink href="#h2">h2</RouterLink><ul><li><RouterLink href="#h3">h3</RouterLink></li></ul></li></ul></nav>
<h1>h1</h1>
<h2>h2</h2>
<h3>h3</h3>
<h4>h4</h4>
<h5>h5</h5>
<h6>h6</h6>
"
`;

exports[`should render toc with default option (h2, h3) > complexTree 1`] = `
"<nav class="table-of-contents"><ul><li><a href="#s1-1">s1-1</a><ul><li><a href="#s1-1-1">s1-1-1</a></li><li><a href="#s1-1-2">s1-1-2</a></li><li><a href="#s1-1-3">s1-1-3</a></li></ul></li><li><a href="#s1-2">s1-2</a></li><li><a href="#s1-3">s1-3</a><ul><li><a href="#s1-3-2">s1-3-2</a></li></ul></li></ul></nav>
<h1>s1</h1>
<h2>s1-1</h2>
<h3>s1-1-1</h3>
<h4>s1-1-1-1</h4>
<h3>s1-1-2</h3>
<h3>s1-1-3</h3>
<h4>s1-1-3-1</h4>
<h2>s1-2</h2>
<h2>s1-3</h2>
<h3>s1-3-2</h3>
<h4>s1-3-2-1</h4>
<h5>s1-3-2-1-1</h5>
<h5>s1-3-2-1-2</h5>
"
`;

exports[`should render toc with default option (h2, h3) > reversedTree 1`] = `
"<nav class="table-of-contents"><ul><li><a href="#h3">h3</a></li><li><a href="#h2">h2</a></li></ul></nav>
<h6>h6</h6>
<h5>h5</h5>
<h4>h4</h4>
<h3>h3</h3>
<h2>h2</h2>
<h1>h1</h1>
"
`;

exports[`should render toc with default option (h2, h3) > simpleTree 1`] = `
"<nav class="table-of-contents"><ul><li><a href="#h2">h2</a><ul><li><a href="#h3">h3</a></li></ul></li></ul></nav>
<h1>h1</h1>
<h2>h2</h2>
<h3>h3</h3>
<h4>h4</h4>
<h5>h5</h5>
<h6>h6</h6>
"
`;
